src/shape.cpp

```diff
@@ -0,0 +1,295 @@
+//*****************************************************************************
+// Copyright 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "shape.hpp"
+
+#include <algorithm>
+#include <exception>
+#include <sstream>
+
+#include "logging.hpp"
+#include "stringutils.hpp"
+
+namespace ovms {
+
+Dimension::Dimension() :
+    Dimension(DYNAMIC_DIMENSION) {
+}
+
+Dimension::Dimension(dimension_value_t dim) :
+    Dimension(dim, dim) {
+}
+
+Dimension::Dimension(dimension_value_t minimum, dimension_value_t maximum) {
+    this->minimum = minimum;
+    this->maximum = maximum;
+}
+
+bool Dimension::isDynamic() const {
+    if (this->minimum != this->maximum)
+        return true;
+    if (this->minimum == DYNAMIC_DIMENSION)
+        return true;
+    return false;
+}
+
+bool Dimension::isStatic() const {
+    return !this->isDynamic();
+}
+
+dimension_value_t Dimension::getAnyValue() const {
+    return this->maximum;
+}
+
+dimension_value_t Dimension::getStaticValue() const {
+    if (this->isDynamic())
+        throw std::invalid_argument("getStaticValue but dimension dynamic");
+    return this->maximum;
+}
+
+dimension_value_t Dimension::getMinValue() const {
+    if (this->isStatic())
+        throw std::invalid_argument("getMinValue but dimension static");
+    return this->minimum;
+}
+
+dimension_value_t Dimension::getMaxValue() const {
+    if (this->isStatic())
+        throw std::invalid_argument("getMaxValue but dimension static");
+    return this->maximum;
+}
+
+bool Dimension::operator==(const Dimension& rhs) const {
+    return this->minimum == rhs.minimum && this->maximum == rhs.maximum;
+}
+
+bool Dimension::operator!=(const Dimension& rhs) const {
+    return !(this->operator==(rhs));
+}
+
+std::string Dimension::toString() const {
+    std::stringstream ss;
+
+    if (this->isStatic()) {
+        ss << this->minimum;
+    } else {
+        if (this->maximum == DYNAMIC_DIMENSION) {
+            ss << DYNAMIC_DIMENSION;
+        } else {
+            ss << "[" << this->minimum << "~" << this->maximum << "]";
```

Comment:
Shouldn't it be ":"?

Reply:
I would insist in leaving it like that, it is simply easier to read